db/migrations/20201220213114_create_circle_basic_schema.js

```diff
@@ -0,0 +1,217 @@
+const { baseDown } = require('../utils')
+
+exports.up = async (knex) => {
+  // circle table
+  let table = 'circle'
+  await knex('entity_type').insert({ table })
+  await knex.schema.createTable(table, (t) => {
+    t.bigIncrements('id').primary()
+    t.string('name').notNullable().unique()
+    t.bigInteger('cover').unsigned()
+    t.bigInteger('avatar').unsigned()
+    t.enu('state', ['active', 'banned', 'archived'])
+      .notNullable()
+      .defaultTo('active')
+    t.bigInteger('owner').unsigned().notNullable()
+    t.string('display_name').notNullable()
+    t.text('description')
+    t.enu('provider', ['stripe']).notNullable().defaultTo('stripe')
+    t.string('provider_product_id').notNullable().unique()
+    t.timestamp('created_at').defaultTo(knex.fn.now())
+    t.timestamp('updated_at').defaultTo(knex.fn.now())
+
+    // foreign keys
+    t.foreign('owner')
+      .references('id')
+      .inTable('user')
+      .onUpdate('CASCADE')
+      .onDelete('CASCADE')
+    t.foreign('cover')
+      .references('id')
+      .inTable('asset')
+      .onUpdate('CASCADE')
+      .onDelete('RESTRICT')
+    t.foreign('avatar')
+      .references('id')
+      .inTable('asset')
+      .onUpdate('CASCADE')
+      .onDelete('RESTRICT')
+  })
+
+  // table for actions on circle, e.g. follow
+  table = 'action_circle'
+
+  await knex('entity_type').insert({ table })
+  await knex.schema.createTable(table, (t) => {
+    t.bigIncrements('id')
+    t.enu('action', ['follow']).defaultTo('follow')
+    t.bigInteger('user_id').unsigned().notNullable()
+    t.bigInteger('target_id').unsigned().notNullable()
+    t.timestamp('created_at').defaultTo(knex.fn.now())
+    t.timestamp('updated_at').defaultTo(knex.fn.now())
+
+    // foreign keys
+    t.foreign('user_id')
+      .references('id')
+      .inTable('user')
+      .onUpdate('CASCADE')
+      .onDelete('CASCADE')
+    t.foreign('target_id')
+      .references('id')
+      .inTable('circle')
+      .onUpdate('CASCADE')
+      .onDelete('CASCADE')
+  })
+
+  // article_circle table
+  table = 'circle_price'
+
+  await knex('entity_type').insert({ table })
+  await knex.schema.createTable(table, (t) => {
+    t.bigIncrements('id')
+
+    t.string('name').notNullable().unique()
+    t.integer('amount').notNullable()
+    t.enu('state', ['active', 'banned', 'archived'])
+    t.enu('currency', ['HKD', 'LIKE']).notNullable().defaultTo('HKD')
+    t.bigInteger('circle_id').unsigned().notNullable()
```

Comment:
I'm thinking that it should use `target_id` & `target_type_id` instead and cut the `circle_` prefix from table name if it isn't just for circle. Same for other `circle_` tables.

cc @devformatters

Reply:
Since for now we can only think of the use case of circle, I think we can reference circle table specifically, and do a migration similar to that of comment table when we need so. Especially when we are not sure how to guarantee referential integrity with `target_id` & `target_type_id`.